source/common/router/config_impl.cc

```diff
@@ -1267,7 +1270,7 @@ RouteConstSharedPtr RouteEntryImplBase::pickWeightedCluster(const Http::HeaderMa
     if (selected_value >= begin && selected_value < end) {
       if (!cluster->clusterHeaderName().get().empty() &&
           !headers.get(cluster->clusterHeaderName()).empty()) {
-        return pickClusterViaClusterHeader(cluster->clusterHeaderName(), headers);
+        return pickClusterViaClusterHeader(cluster->clusterHeaderName(), headers, cluster.get());
```

Comment:
static_cast `cluster.get()` to `RouteEntryAndRoute*`? absl::implicit_cast should be best option here but I guess it is not available in Envoy code base.